src/stacked_borrows/diagnostics.rs

```diff
@@ -342,32 +343,39 @@ impl<'span, 'history, 'ecx, 'mir, 'tcx> DiagnosticCx<'span, 'history, 'ecx, 'mir
 
         let protected = protector_tag
             .and_then(|protector| {
-                self.history.protectors.iter().find_map(|protection| {
-                    if protection.tag == protector {
-                        Some((protection.orig_tag, protection.span.data()))
-                    } else {
-                        None
-                    }
+                self.history.protectors.iter().find(|protection| {
+                    protection.tag == protector
                 })
             })
-            .and_then(|(tag, call_span)| {
+            .and_then(|protection| {
                 self.history.creations.iter().rev().find_map(|event| {
-                    if ProvenanceExtra::Concrete(event.retag.new_tag) == tag {
-                        Some((event.retag.orig_tag, event.span.data(), call_span))
+                    if ProvenanceExtra::Concrete(event.retag.new_tag) == protection.orig_tag {
+                        Some((protection, event))
                     } else {
                         None
                     }
                 })
             })
-            .map(|(protecting_tag, protecting_tag_span, protection_span)| {
+            .map(|(protection, protection_parent)| {
+                let protected_tag = protection.tag;
                 [
                     (
                         format!(
-                            "{tag:?} was protected due to {protecting_tag:?} which was created here"
+                            "{tag:?} cannot be used for memory access because that would remove protected tag {protected_tag:?}, protected by this function call",
                         ),
-                        protecting_tag_span,
+                        protection.span.data(),
                     ),
-                    (format!("this protector is live for this call"), protection_span),
+                    if protection_parent.retag.new_tag == tag {
+                        (format!("{protected_tag:?} was derived from {tag:?}, the tag used for this memory access"), DUMMY_SP.data())
+                    } else {
+                        (
+                            format!(
+                                "{protected_tag:?} was derived from {protected_parent_tag:?}, which in turn was created here",
```

Comment:
In `aliasing_mut6.rs`, we have this error:
```
error: Undefined Behavior: not granting access to tag <3095> because incompatible item [Unique for <3099>] is protected by call 943
  --> tests/fail/stacked_borrows/illegal_write6.rs:10:14
   |
10 |     unsafe { *y = 2 }; //~ ERROR: /not granting access .* because incompatible item .* is protected/
   |              ^^^^^^ not granting access to tag <3095> because incompatible item [Unique for <3099>] is protected by call 943
   |
```
...

```
help: <3099> was derived from <3096>, which in turn was created here
  --> tests/fail/stacked_borrows/illegal_write6.rs:4:9
   |
4  |     foo(x, p);
   |         ^
```
So I'm confused (again), because this message suggests that this is the span where `<3096>` was created, but this is the function call, where we should be creating `<3099>`. Or is this me reaching for a parent retag in order to get a span which is outside the function (because the spands of FnEntry retags aren't helpful)?

Reply:
`<3099>` is created inside `foo`, just after it got called. `<3096>` is created in the caller; that use of `c` is an implicit reborrow which creates a new tag.